hoot-core/src/main/cpp/hoot/core/io/ApiDbReader.cpp

```diff
@@ -402,19 +404,110 @@ void ApiDbReader::_readByBounds(OsmMapPtr map, const Envelope& bounds)
         }
       }
       LOG_VARD(relationIds.size());
-
-      if (relationIds.size() > 0)
+      //  Iterate all relations (and sub-relations) that are "within" the bounds
+      QSet<QString> completedRelationIds;
+      while (relationIds.size() > 0)
       {
+        QSet<QString> newNodes;
+        QSet<QString> newWays;
+        QSet<QString> newRelations;
         LOG_DEBUG("Retrieving relations by relation ID...");
         boost::shared_ptr<QSqlQuery> relationItr =
           _getDatabase()->selectElementsByElementIdList(relationIds, TableType::Relation);
         while (relationItr->next())
         {
           RelationPtr relation = _resultToRelation(*relationItr, *map);
           map->addElement(relation);
+          const vector<RelationData::Entry>& members = relation->getMembers();
+          //  Iterate all members so that they can be retrieved later
+          for (vector<RelationData::Entry>::const_iterator it = members.begin(); it != members.end(); ++it)
+          {
+            ElementType type = it->getElementId().getType();
+            QString id = QString::number(it->getElementId().getId());
+            if (type == ElementType::Node)
+              newNodes.insert(id);
+            else if (type == ElementType::Way)
+              newWays.insert(id);
+            else if (type == ElementType::Relation)
+              newRelations.insert(id);
+            else
```

Comment:
Here is where the actual "hydration" of the relations happens, the new nodes/ways/relations that are members of the relations already retrieved are noted and then queried below.